### PR TITLE
Move core logic out of UTF into its own package

### PR DESCRIPTION
**Pull request by lewijacn**
Signed-off-by: Tanner Lewis <lewijacn@amazon.com>

### Description
This change splits the core functionality of the UTF package (clients, step-wise framework runner, cluster manipulation logic) into its own package. This will give us greater flexibility in the future with allowing other packages to have this common functionality they will likely need, as well as eliminate a lot of code duplication that could arise such as common rest clients needed by both the UTF and robot tests.

Also included were some minor changes such as Lint fixes, adjusting github action unit tests, etc.

As a side note, the way the UTF had been structured and segregated made it very easy to pull this logic out

### Issues Resolved
* https://github.com/opensearch-project/opensearch-migrations/issues/61

### Testing
Cluster migration core unit tests
```
(.venv) (10:48:37) local [/Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/cluster_migration_core] -> python -m pytest unit_tests/
================================================================================================================================================= test session starts ==================================================================================================================================================
platform darwin -- Python 3.10.7, pytest-7.2.0, pluggy-1.0.0
rootdir: /Volumes/workplace/opensearch/lewijacn-migrations/opensearch-migrations/cluster_migration_core
collected 104 items                                                                                                                                                                                                                                                                                                    

unit_tests/clients/test_rest_client_default.py .........                                                                                                                                                                                                                                                         [  8%]
unit_tests/clients/test_rest_ops.py ....                                                                                                                                                                                                                                                                         [ 12%]
unit_tests/cluster_management/test_cluster.py ............                                                                                                                                                                                                                                                       [ 24%]
unit_tests/cluster_management/test_container_configuration.py .                                                                                                                                                                                                                                                  [ 25%]
unit_tests/cluster_management/test_docker_framework_client.py ...............                                                                                                                                                                                                                                    [ 39%]
unit_tests/cluster_management/test_node.py ...........                                                                                                                                                                                                                                                           [ 50%]
unit_tests/cluster_management/test_node_configuration.py ..                                                                                                                                                                                                                                                      [ 51%]
unit_tests/core/test_exception_base.py ...                                                                                                                                                                                                                                                                       [ 54%]
unit_tests/core/test_expectation.py .........                                                                                                                                                                                                                                                                    [ 63%]
unit_tests/core/test_framework_runner.py ......                                                                                                                                                                                                                                                                  [ 69%]
unit_tests/core/test_framework_state.py ...                                                                                                                                                                                                                                                                      [ 72%]
unit_tests/core/test_framework_step.py ....                                                                                                                                                                                                                                                                      [ 75%]
unit_tests/core/test_logging_wrangler.py .                                                                                                                                                                                                                                                                       [ 76%]
unit_tests/core/test_shell_interactions.py .......                                                                                                                                                                                                                                                               [ 83%]
unit_tests/core/test_test_config_wrangling.py .........                                                                                                                                                                                                                                                          [ 92%]
unit_tests/core/test_versions_engine.py .......                                                                                                                                                                                                                                                                  [ 99%]
unit_tests/core/test_workspace_wrangler.py .                                                                                                                                                                                                                                                                     [100%]

================================================================================================================================================= 104 passed in 0.79s ==================================================================================================================================================
```

UTF snapshot restore test run
```
(.venv) (10:49:13) local [~/workplace/opensearch/lewijacn-migrations/opensearch-migrations/upgrades] -> ./run_utf.py --test_config test_config.json
[FrameworkRunner] ============ Running Step: LoadTestConfig ============
[LoadTestConfig] Loading test config file...
[LoadTestConfig] Loaded test config file successfully
[FrameworkRunner] Step Succeeded: LoadTestConfig
[FrameworkRunner] ============ Running Step: BootstrapDocker ============
[BootstrapDocker] Checking if Docker is installed and available...
[BootstrapDocker] Docker appears to be installed and available
[BootstrapDocker] Ensuring the Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available either locally or remotely...
[BootstrapDocker] Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available
[BootstrapDocker] Ensuring the Docker image opensearchproject/opensearch:1.3.6 is available either locally or remotely...
[BootstrapDocker] Docker image opensearchproject/opensearch:1.3.6 is available
[FrameworkRunner] Step Succeeded: BootstrapDocker
[FrameworkRunner] ============ Running Step: SnapshotRestoreSetup ============
[SnapshotRestoreSetup] Creating shared Docker volume to share snapshots...
[SnapshotRestoreSetup] Created shared Docker volume cluster-snapshots-volume
[FrameworkRunner] Step Succeeded: SnapshotRestoreSetup
[FrameworkRunner] ============ Running Step: SelectExpectations ============
[SelectExpectations] Loading Knowledge Base and filtering for relevant expectations
[SelectExpectations] Found 2 relevant expectations.
[FrameworkRunner] Step Succeeded: SelectExpectations
[FrameworkRunner] ============ Running Step: StartSourceCluster ============
[StartSourceCluster] Creating source cluster...
[StartSourceCluster] Waiting up to 30 sec for cluster to be active...
Node source-cluster-node-1 is now active
Node source-cluster-node-2 is now active
[StartSourceCluster] Cluster source-cluster is active
[FrameworkRunner] Step Succeeded: StartSourceCluster
[FrameworkRunner] ============ Running Step: PerformPreUpgradeTest ============
==============================================================================
Tests                                                                         
==============================================================================
Tests.Common Upgrade Test                                                     
==============================================================================
Perform pre-upgrade setup of "consistent-document-count" expectation  PUT http://localhost:9200/sample-index [status:200 request:0.462s]
.POST http://localhost:9200/sample-index/_doc [status:201 request:0.173s]
.POST http://localhost:9200/sample-index/_doc [status:201 request:0.009s]
.POST http://localhost:9200/sample-index/_refresh [status:200 request:0.038s]
.POST http://localhost:9200/sample-index/_count [status:200 request:0.049s]
Perform pre-upgrade setup of "consistent-document-count" expectation  | PASS |
------------------------------------------------------------------------------
Tests.Common Upgrade Test                                             | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /tmp/utf/test-results/pre_upgrade/output.xml
Log:     /tmp/utf/test-results/pre_upgrade/log.html
Report:  /tmp/utf/test-results/pre_upgrade/report.html
[FrameworkRunner] Step Succeeded: PerformPreUpgradeTest
[FrameworkRunner] ============ Running Step: CreateSourceSnapshot ============
[CreateSourceSnapshot] Creating snapshot of source cluster...
[CreateSourceSnapshot] Confirming snapshot of source cluster exists...
[CreateSourceSnapshot] Source snapshot created successfully
[FrameworkRunner] Step Succeeded: CreateSourceSnapshot
[FrameworkRunner] ============ Running Step: StartTargetCluster ============
[StartTargetCluster] Creating target cluster...
[StartTargetCluster] Waiting up to 30 sec for cluster to be active...
Node target-cluster-node-1 is now active
Node target-cluster-node-2 is now active
[StartTargetCluster] Cluster target-cluster is active
[FrameworkRunner] Step Succeeded: StartTargetCluster
[FrameworkRunner] ============ Running Step: RestoreSourceSnapshot ============
[RestoreSourceSnapshot] Checking if source snapshots are visible on target...
[RestoreSourceSnapshot] Source snapshot visible to target cluster
[RestoreSourceSnapshot] Restoring source snapshot onto target...
[RestoreSourceSnapshot] Snapshot restored successfully
[FrameworkRunner] Step Succeeded: RestoreSourceSnapshot
[FrameworkRunner] ============ Running Step: PerformPostUpgradeTest ============
==============================================================================
Tests                                                                         
==============================================================================
Tests.Common Upgrade Test                                                     
==============================================================================
Perform post-upgrade assertion of "consistent-document-count" expe... .POST http://localhost:9202/sample-index/_count [status:200 request:0.101s]
Perform post-upgrade assertion of "consistent-document-count" expe... | PASS |
------------------------------------------------------------------------------
Tests.Common Upgrade Test                                             | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /tmp/utf/test-results/post_upgrade/output.xml
Log:     /tmp/utf/test-results/post_upgrade/log.html
Report:  /tmp/utf/test-results/post_upgrade/report.html
[FrameworkRunner] Step Succeeded: PerformPostUpgradeTest
[FrameworkRunner] ============ Running Step: StopSourceCluster ============
[StopSourceCluster] Stopping cluster source-cluster...
[StopSourceCluster] Cleaning up underlying resources for cluster source-cluster...
[FrameworkRunner] Step Succeeded: StopSourceCluster
[FrameworkRunner] ============ Running Step: StopTargetCluster ============
[StopTargetCluster] Stopping cluster target-cluster...
[StopTargetCluster] Cleaning up underlying resources for cluster target-cluster...
[FrameworkRunner] Step Succeeded: StopTargetCluster
[FrameworkRunner] ============ Running Step: SnapshotRestoreTeardown ============
[SnapshotRestoreTeardown] Removing shared Docker volume cluster-snapshots-volume...
[SnapshotRestoreTeardown] Removed shared Docker volume cluster-snapshots-volume
[FrameworkRunner] Step Succeeded: SnapshotRestoreTeardown
[FrameworkRunner] ============ Running Step: ReportResults ============
[ReportResults] ======================== SKIPPED EXPECTATIONS =======================
[ReportResults] doy-format-date-range-query-bug
[ReportResults] 	Description: Given 1/ an index created in ES 7.x; 2/ documents with dates uploaded in a yyyy-DDD (day of year) format,
[ReportResults] 	and a query that uses an inclusive date range, the results of the query will fail to include documents where the date
[ReportResults] 	of the document is the same day as the inclusive endpoint of the range. See https://github.com/opensearch-
[ReportResults] 	project/OpenSearch/issues/4285
[ReportResults] 

[ReportResults] ======================== PASSED EXPECTATIONS ========================
[ReportResults] consistent-document-count
[ReportResults] 	Phases: pre_upgrade, post_upgrade
[ReportResults] 

[ReportResults] For additional details on test results, see the test result directory: /tmp/utf/test-results
[FrameworkRunner] Step Succeeded: ReportResults
[FrameworkRunner] Ran through all 14 steps successfully
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2023-01-04_10_49_18
```

### Check List
- [ ] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: cluster_migration_core/unit_tests/core/test_framework_runner.py [hunk 1]
```diff
@@ -97,8 +97,6 @@ def state_file_for_resume(test_workspace):
 TEST_CONFIG_FILE_PATH = "./path/to/test_config.json"
 
 
-# This patch replaces the namespace of the "steps" module with that of this current file
-@mock.patch('upgrade_testing_framework.core.framework_runner.steps', sys.modules[__name__])
```
Comment (lewijacn) [inline cluster_migration_core/unit_tests/core/test_framework_runner.py:101]: @chelma I wasn't able to determine how this was being used as it didn't seem to affect test cases, lmk if I am missing something here

Comment (chelma) [reply cluster_migration_core/unit_tests/core/test_framework_runner.py:101]: Interesting.  Hypothetically, you just need to swap `upgrade_testing_framework` for `cluster_migration_core`, but I'm confused why it didn't cause test failures.  Can you either perform that swap or delete the line of code?

Comment (lewijacn) [reply cluster_migration_core/unit_tests/core/test_framework_runner.py:101]: This is removed currently and no issue I can see, just curious if we still actually needed it or not. The framework runner originally had an import for steps with a comment that it was only used for unit tests, which I removed as steps now belongs in UTF so this patch was failing.

**Review by chelma — APPROVED** [review]
LGTM - Thanks!

File: cluster_migration_core/unit_tests/core/test_framework_runner.py (hunk 1 above)
Comment (chelma) [reply cluster_migration_core/unit_tests/core/test_framework_runner.py:101]: I guess we didn't need it.  It was something I copied/pasted from my tests in the Vingilot codebase.

**Comment by okhasawn** [comment]
I see a "cluster_migration_core" folder within a "cluster_migration_core" folder, do you think we need to rename one maybe?

**Review by okhasawn — APPROVED** [review]

**Pull request merged**